src/meta/test/meta_bulk_load_service_test.cpp

```diff
@@ -956,6 +966,56 @@ TEST_F(bulk_load_failover_test, downloading_with_partition_wrong_status)
     ASSERT_FALSE(app_is_bulk_loading(SYNC_APP_NAME));
 }
 
+// app:downloading, partition[0~3]=downloading
```

Comment:
These five tests are mostly in common. You can reduce the repeated code by:

```cpp
TEST_F(bulk_load_failover_test, failover_states) {
  struct test_data {
    std::function<void()> mock_current_state_f;
  } tests[] = {
    // app:downloading, partition[0~3]=downloading
    []() { mock_pstatus_map(bulk_load_status::BLS_DOWNLOADING, SYNC_PARTITION_COUNT - 1); }

    // app:downloading, partition[0~3]=downloaded
    []() { mock_pstatus_map(bulk_load_status::BLS_DOWNLOADED, SYNC_PARTITION_COUNT - 1); }

    // app:downloading, partition[0]=downloaded, partition[1~3]=downloading
    [this]() {
      _pstatus_map[0] = bulk_load_status::BLS_DOWNLOADED;
      mock_pstatus_map(bulk_load_status::BLS_DOWNLOADING, SYNC_PARTITION_COUNT - 1, 1);
    },
  };
  for (const auto& tt : tests) {
    Setup();
    tt.mock_current_state_f();
    try_to_continue_bulk_load(bulk_load_status::BLS_DOWNLOADING);
    ASSERT_EQ(get_app_bulk_load_status(SYNC_APP_ID), bulk_load_status::BLS_DOWNLOADING);
    ASSERT_EQ(get_app_in_process_count(SYNC_APP_ID), SYNC_PARTITION_COUNT);
    ASSERT_TRUE(app_is_bulk_loading(SYNC_APP_NAME));
    Teardown();
  }
```

Btw, just curious, why there could be a state transition from BLS_DOWNLOADED to BLS_DOWNLOADING? This is in your second test.